presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -358,7 +361,52 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
         return new RelationPlan(root, analysis.getScope(node), outputSymbols);
     }
 
-    private static boolean isEqualComparisonExpression(Expression conjunct)
+    private Optional<Unnest> getUnnest(Relation relation)
+    {
+        if (relation instanceof AliasedRelation) {
+            return getUnnest(((AliasedRelation) relation).getRelation());
+        }
+        if (relation instanceof Unnest) {
+            return Optional.of((Unnest) relation);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<Lateral> getLateral(Relation relation)
+    {
+        if (relation instanceof AliasedRelation) {
+            return getLateral(((AliasedRelation) relation).getRelation());
+        }
+        if (relation instanceof Lateral) {
+            return Optional.of((Lateral) relation);
+        }
+        return Optional.empty();
+    }
+
+    private RelationPlan planLateralJoin(Node joinNode, RelationPlan leftPlan, Lateral lateral)
+    {
+        RelationPlan rightPlan = process(lateral.getQuery(), null);
+        PlanBuilder leftPlanBuilder = initializePlanBuilder(leftPlan);
+        PlanBuilder planBuilder = subqueryPlanner.appendApplyNode(
```

Comment:
>I would restore the previous version of ApplyNode as RelationalApplyNode (it can be though as LateralJoin node) and used it for LATERAL join

Are there any other possible uses of such `RelationalApplyNode` other than lateral joins? Also you would need to keep information if `RelationalApplyNode` is `LEFT` or `INNER`. Those are usually join node properties. On the other hand everybody would know what `LateralJoinNode`  is since it has well defined meaning (the same applies to `SemiJoinNode`).

> Additionally, I would also rename current as ScalarApplyNode. That way, we could use RelationalApplyNode for all the current ApplyNode#isResolved...()

`ScalarApplyNode` name might be confusing with scalar or non-scalar subquery since `ScalarApplyNode` would support both.
Would you like `RelationalApplyNode` (aka `LateralJoinNode`) to be supertype of `ScalarApplyNode`? It seems that those two nodes have little in common apart form correlation support. Optimizers that transform `ScalarApplyNode` won't be executed for `RelationalApplyNode` and vice versa. There is some common code (e.g: `extractCorrelation` or `PlanNodeDecorrelator`), but it is/can be extracted to separate methods or classes.

Anyway I would need to see the code. If things simplify and are more readable that would be a big plus in this hard topic.